command/monitor/monitor.go

```diff
@@ -0,0 +1,148 @@
+package monitor
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+// Monitor provides a mechanism to stream logs using go-hclog
+// InterceptLogger and SinkAdapter. It allows streaming of logs
+// at a different log level than what is set on the logger.
+type Monitor interface {
+	// Start returns a channel of log messages which are sent
+	// ever time a log message occurs
+	Start() <-chan []byte
+
+	// Stop de-registers the sink from the InterceptLogger
+	// and closes the log channels
+	Stop()
+}
+
+// monitor implements the Monitor interface
+type monitor struct {
+	sink log.SinkAdapter
+
+	// logger is the logger we will be monitoring
+	logger log.InterceptLogger
+
+	// logCh is a buffered chan where we send logs when streaming
+	logCh chan []byte
+
+	// doneCh coordinates the shutdown of logCh
+	doneCh chan struct{}
+
+	// droppedCount is the current count of messages
+	// that were dropped from the logCh buffer.
+	droppedCount uint64
+	bufSize      int
+	// dropCheckInterval is the amount of time we should
+	// wait to check for dropped messages. Defaults
+	// to 3 seconds
+	dropCheckInterval time.Duration
+}
+
+// New creates a new Monitor. Start must be called in order to actually start
+// streaming logs
+func NewMonitor(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) Monitor {
```

Comment:
I recall this was discussed in a previous conversation, but since `monitor` is the package name we could simply call this `New` (similar to `errors.New`). The internal method that returns the concrete type could be called something like `newBufferedMonitor` (to avoid overloading `new` and to provide a bit more context on the type returned), with `monitor` being renamed to `bufferedMonitor`.

Reply:
That's how he had it until I talked him out of it.  I think it's a bad idea to use a constructor called `New`, because as soon as you introduce something else in the same package that needs a constructor, now you have an inconsistency in your naming.  It also is no fun to grep for.

Reply:
Fair enough. I thought there was an idiomatic way for constructor naming, but I do see both practices in the stdlib.